plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,410 @@
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		if (!fu_udev_device_pwrite(FU_UDEV_DEVICE(self),
+					   fu_chunk_get_address(chk),
+					   fu_chunk_get_data(chk),
+					   fu_chunk_get_data_sz(chk),
+					   error)) {
+			g_prefix_error(error,
+				       "failed to write @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_erase(FuAndroidBootDevice *self, FuProgress *progress, GError **error)
+{
+	g_autoptr(GPtrArray) chunks = NULL;
+	gsize bufsz = fu_device_get_firmware_size_max(FU_DEVICE(self));
+	g_autofree guint8 *buf = g_malloc0(bufsz);
+
+	chunks = fu_chunk_array_new(buf, bufsz, 0x0, 0x0, 10 * 1024);
+
+	if (g_getenv("FWUPD_ANDROID_BOOT_VERBOSE") != NULL)
+		fu_dump_bytes(G_LOG_DOMAIN, "erase", g_bytes_new(buf, bufsz));
+
+	if (!fu_android_boot_device_write(self, chunks, progress, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_verify(FuAndroidBootDevice *self,
+			      GPtrArray *chunks,
+			      FuProgress *progress,
+			      GError **error)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+
+	/* verify each chunk */
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autofree guint8 *buf = g_malloc0(fu_chunk_get_data_sz(chk));
+		g_autoptr(GBytes) blob1 = fu_chunk_get_bytes(chk);
+		g_autoptr(GBytes) blob2 = NULL;
+
+		if (!fu_udev_device_pread(FU_UDEV_DEVICE(self),
+					  fu_chunk_get_address(chk),
+					  buf,
+					  fu_chunk_get_data_sz(chk),
+					  error)) {
+			g_prefix_error(error,
+				       "failed to read @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		blob2 = g_bytes_new_static(buf, fu_chunk_get_data_sz(chk));
+		if (!fu_bytes_compare(blob1, blob2, error)) {
+			g_prefix_error(error,
+				       "failed to verify @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* get data to write */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	if (g_getenv("FWUPD_ANDROID_BOOT_VERBOSE") != NULL)
+		fu_dump_bytes(G_LOG_DOMAIN, "write", fw);
+
+	if (g_bytes_get_size(fw) > fu_device_get_firmware_size_max(device)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "firmware too large, got 0x%x, expected <= 0x%x",
+			    (guint)g_bytes_get_size(fw),
+			    (guint)fu_device_get_firmware_size_max(device));
+		return FALSE;
+	}
```

Comment:
doesn't fu_device_write_firmware do exactly this check? again coffee...

Reply:
I have seen this in other places like the MTD plugin as well. That's why I have it. Not sure either.

Reply:
If the MDT plugin implements ->prepare_firmware() then it indeed needs to do it, otherwise fu_device_prepare_firmware() does it for you.